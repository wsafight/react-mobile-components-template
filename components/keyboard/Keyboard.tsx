import React, { PureComponent } from 'react';
import classnames from 'classnames';
import PropsType from './PropsType';
import Icon from '../icon';

const KEYS = {
  number: ['1', '2', '3', '4', '5', '6', '7', '8', '9', '', '0', 'close'],
  price: ['1', '2', '3', '4', '5', '6', '7', '8', '9', '.', '0', 'close'],
  idcard: ['1', '2', '3', '4', '5', '6', '7', '8', '9', 'X', '0', 'close'],
};

export interface KeyboardProps extends PropsType {
  prefixCls?: string;
  className?: string;
}

export default class Keyboard extends PureComponent<KeyboardProps, {}> {
  static defaultProps: KeyboardProps = {
    prefixCls: 'za-keyboard',
    type: 'number',
  };

  private longPressTimer;

  onLongPressIn = (key: string) => {
    this.onKeyClick(key);
    this.longPressTimer = setTimeout(() => {
      this.longPressTimer = setInterval(() => {
        this.onKeyClick(key);
      }, 100);
    }, 800);
  };

  onLongPressOut = (e) => {
    e.preventDefault();
    clearInterval(this.longPressTimer);
  };

  onKeyClick = (key: string) => {
    if (key.length === 0) {
      return;
    }

    const { onKeyClick } = this.props;
    if (typeof onKeyClick === 'function') {
      onKeyClick(key);
    }
  };

  getKeys = () => {
    const { type } = this.props;
    return type ? KEYS[type] : KEYS.number;
  };

  renderKey = (text: string, index: number) => {
    const { prefixCls } = this.props;

    const keyCls = classnames(`${prefixCls}__item`, {
      [`${prefixCls}__item--disabled`]: text.length === 0,
    });

    return (
      <div className={keyCls} key={+index} onClick={() => this.onKeyClick(text)}>
        {text === 'close' ? <Icon type="keyboard" size="lg" /> : text}
      </div>
    );
  };

  render() {
    const { prefixCls, className, locale } = this.props;
    const cls = classnames(prefixCls, className);

    return (
      <div className={cls}>
        <div className={`${prefixCls}__keys`}>{this.getKeys().map(this.renderKey)}</div>
        <div className={`${prefixCls}__handle`}>
          <div
            className={`${prefixCls}__item`}
            onTouchStart={() => this.onLongPressIn('delete')}
            onTouchEnd={this.onLongPressOut}
            onTouchCancel={this.onLongPressOut}
            onMouseDown={() => this.onLongPressIn('delete')}
            onMouseUp={this.onLongPressOut}
          >
            <Icon type="deletekey" size="lg" />
          </div>
          <div
            className={`${prefixCls}__item ${prefixCls}__item--ok`}
            onClick={() => this.onKeyClick('ok')}
          >
            {locale!.okText}
          </div>
        </div>
      </div>
    );
  }
}
