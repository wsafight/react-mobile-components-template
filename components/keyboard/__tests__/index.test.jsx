import React from 'react';
import { render, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import Keyboard from '../index';

describe('Keyboard', () => {
  it('renders correctly', () => {
    const wrapper = render(<Keyboard />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('type is price', () => {
    const wrapper = render(<Keyboard type="price" />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('type is idcard', () => {
    const wrapper = render(<Keyboard type="idcard" />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('click keyboard', () => {
    const onKeyClick = jest.fn();
    const wrapper = mount(<Keyboard onKeyClick={onKeyClick} />);
    const keys = wrapper.find('.za-keyboard__keys');
    expect(keys.childAt(0).text()).toBe('1');
    keys.childAt(0).simulate('click');
    expect(onKeyClick).toBeCalledWith('1');
    keys.childAt(9).simulate('click');
    keys.childAt(11).simulate('click');
    expect(onKeyClick).toBeCalledWith('close');

    const handles = wrapper.find('.za-keyboard__handle');
    handles.childAt(0).simulate('touchStart');
    expect(onKeyClick).toBeCalledWith('delete');
    handles.childAt(1).simulate('click');
    expect(onKeyClick).toBeCalledWith('ok');
  });
});
