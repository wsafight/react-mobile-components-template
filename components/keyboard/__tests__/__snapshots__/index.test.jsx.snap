// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Keyboard renders correctly 1`] = `
<div
  class="za-keyboard"
>
  <div
    class="za-keyboard__keys"
  >
    <div
      class="za-keyboard__item"
    >
      1
    </div>
    <div
      class="za-keyboard__item"
    >
      2
    </div>
    <div
      class="za-keyboard__item"
    >
      3
    </div>
    <div
      class="za-keyboard__item"
    >
      4
    </div>
    <div
      class="za-keyboard__item"
    >
      5
    </div>
    <div
      class="za-keyboard__item"
    >
      6
    </div>
    <div
      class="za-keyboard__item"
    >
      7
    </div>
    <div
      class="za-keyboard__item"
    >
      8
    </div>
    <div
      class="za-keyboard__item"
    >
      9
    </div>
    <div
      class="za-keyboard__item za-keyboard__item--disabled"
    />
    <div
      class="za-keyboard__item"
    >
      0
    </div>
    <div
      class="za-keyboard__item"
    >
      <i
        class="za-icon za-icon--keyboard za-icon--default za-icon--lg"
        theme="default"
        type="keyboard"
      >
        <svg
          fill="currentColor"
          height="1em"
          viewBox="0 0 32 32"
          width="1em"
        >
          <use
            href="#keyboard"
          />
        </svg>
      </i>
    </div>
  </div>
  <div
    class="za-keyboard__handle"
  >
    <div
      class="za-keyboard__item"
    >
      <i
        class="za-icon za-icon--deletekey za-icon--default za-icon--lg"
        theme="default"
        type="deletekey"
      >
        <svg
          fill="currentColor"
          height="1em"
          viewBox="0 0 32 32"
          width="1em"
        >
          <use
            href="#deletekey"
          />
        </svg>
      </i>
    </div>
    <div
      class="za-keyboard__item za-keyboard__item--ok"
    >
      确定
    </div>
  </div>
</div>
`;

exports[`Keyboard type is idcard 1`] = `
<div
  class="za-keyboard"
>
  <div
    class="za-keyboard__keys"
  >
    <div
      class="za-keyboard__item"
    >
      1
    </div>
    <div
      class="za-keyboard__item"
    >
      2
    </div>
    <div
      class="za-keyboard__item"
    >
      3
    </div>
    <div
      class="za-keyboard__item"
    >
      4
    </div>
    <div
      class="za-keyboard__item"
    >
      5
    </div>
    <div
      class="za-keyboard__item"
    >
      6
    </div>
    <div
      class="za-keyboard__item"
    >
      7
    </div>
    <div
      class="za-keyboard__item"
    >
      8
    </div>
    <div
      class="za-keyboard__item"
    >
      9
    </div>
    <div
      class="za-keyboard__item"
    >
      X
    </div>
    <div
      class="za-keyboard__item"
    >
      0
    </div>
    <div
      class="za-keyboard__item"
    >
      <i
        class="za-icon za-icon--keyboard za-icon--default za-icon--lg"
        theme="default"
        type="keyboard"
      >
        <svg
          fill="currentColor"
          height="1em"
          viewBox="0 0 32 32"
          width="1em"
        >
          <use
            href="#keyboard"
          />
        </svg>
      </i>
    </div>
  </div>
  <div
    class="za-keyboard__handle"
  >
    <div
      class="za-keyboard__item"
    >
      <i
        class="za-icon za-icon--deletekey za-icon--default za-icon--lg"
        theme="default"
        type="deletekey"
      >
        <svg
          fill="currentColor"
          height="1em"
          viewBox="0 0 32 32"
          width="1em"
        >
          <use
            href="#deletekey"
          />
        </svg>
      </i>
    </div>
    <div
      class="za-keyboard__item za-keyboard__item--ok"
    >
      确定
    </div>
  </div>
</div>
`;

exports[`Keyboard type is price 1`] = `
<div
  class="za-keyboard"
>
  <div
    class="za-keyboard__keys"
  >
    <div
      class="za-keyboard__item"
    >
      1
    </div>
    <div
      class="za-keyboard__item"
    >
      2
    </div>
    <div
      class="za-keyboard__item"
    >
      3
    </div>
    <div
      class="za-keyboard__item"
    >
      4
    </div>
    <div
      class="za-keyboard__item"
    >
      5
    </div>
    <div
      class="za-keyboard__item"
    >
      6
    </div>
    <div
      class="za-keyboard__item"
    >
      7
    </div>
    <div
      class="za-keyboard__item"
    >
      8
    </div>
    <div
      class="za-keyboard__item"
    >
      9
    </div>
    <div
      class="za-keyboard__item"
    >
      .
    </div>
    <div
      class="za-keyboard__item"
    >
      0
    </div>
    <div
      class="za-keyboard__item"
    >
      <i
        class="za-icon za-icon--keyboard za-icon--default za-icon--lg"
        theme="default"
        type="keyboard"
      >
        <svg
          fill="currentColor"
          height="1em"
          viewBox="0 0 32 32"
          width="1em"
        >
          <use
            href="#keyboard"
          />
        </svg>
      </i>
    </div>
  </div>
  <div
    class="za-keyboard__handle"
  >
    <div
      class="za-keyboard__item"
    >
      <i
        class="za-icon za-icon--deletekey za-icon--default za-icon--lg"
        theme="default"
        type="deletekey"
      >
        <svg
          fill="currentColor"
          height="1em"
          viewBox="0 0 32 32"
          width="1em"
        >
          <use
            href="#deletekey"
          />
        </svg>
      </i>
    </div>
    <div
      class="za-keyboard__item za-keyboard__item--ok"
    >
      确定
    </div>
  </div>
</div>
`;
